keps/sig-api-machinery/customresources-pruning-and-defaulting.md

```diff
@@ -0,0 +1,510 @@
+
+### Excluding values from Pruning
+
+There are cases where parts of an object are verbatim JSON, i.e. without any applied schema and especially without a complete specification which allows to apply pruning. E.g. the [`JSONSchemaProps` types in apiextensions-apiserver contain fields of type JSON](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types_jsonschema.go#L28) for defaults, examples and enums. These should never be pruned (and they are not pruned while decoding using JSON marshalling).
+
+Hence, we need a mechanism to express that in the OpenAPI validation schema (compare https://github.com/kubernetes/kubernetes/pull/64558#issuecomment-403564033).
+
+**Raw JSON Option 1:** add a format `json`, e.g.:
+
+```json
+{"properties": {"x": {"format": "json"}}}
+```
+
+In this example "x" is excluded from pruning. Note that you can still use any kind of OpenAPI validation schema constructs to restrict `"x"` further.
+
+Note that we lose expressiveness of the existing `format` strings: we either apply the format to `"json"` or any other pre-defined format. I.e. we cannot express that pruning should be disabled, but if it is an integer, it should be an int32.
+
+**Question:** this feels like a reasonable loss of expressivity. Do we accept that?
+
+**Raw JSON Option 2:** add an extension property, e.g.
+
+```json
+{"properties": {"x": {"x-no-pruning": true}}}
+```
+
+This would not lead to a loss in expressivity. It is formulated negatively intentionally to have `false` as the default with `omitempty`.
+
+As OpenAPI extensions are prefixed with `x-kubernetes` elsewhere, we probably should name it `x-kubernetes-no-pruning: true`.
+
+**We propose to follow option with the `x-kubernetes` prefix** because it is more explicit, does not reduce expressivity and does not mangle with the already very vague `format` field definition.
+
+### Nested x-no-pruning
+
+**Question:** should we support nested `x-no-pruning`, i.e. disabling pruning for a sub-object, but re-enable it something deep inside of it? E.g.
+
+```json
+{
+    "properties": {
+        "x": {
+            "x-no-pruning": true,
+            "properties": {
+		        "y": { 
+                    "x-kubernetes-no-pruning": false,
+                    "properties": { "z": {} }
+                }
+            }
+        }
+    }
+}
+```
+
+If we do, the object 
+
+```json
+{
+    "a": 1,
+    "x": {
+        "b": 2,
+        "y": {
+            "c": 3,
+            "z": 42
+        }
+    }
+}
+```
+
+is pruned to  `{"x":{"b": 2, "y":{"z": 42}}}`.
+
+## Opt-in and Opt-out of Pruning on CRD Level
+
+We will add a pruning flag to `CustomResourceDefinitionSpec` of `apiextensions.k8s.io/v1beta1`:
+
+```golang
+type CustomResourceDefinitionSpec struct {
+  ...
+	
+  // Prune enables pruning of unspecified fields. Defaults to false.
```

Comment:
I'm not sure about this. Yes, it's a behavior change. But the people who need it are the people who won't turn this flag on. How do we avoid surprising them?

A tool that you can run that scans your CRs and flips this bit if none of them would get pruned, and sends you a email otherwise? It seems hard.

The only reason not to turn it on is that people might have a spec that's missing an important field. Maybe in 1.12 we can somehow make the CRD's status field indicate whether CRs are stored that need pruning, and then tell people they can't upgrade to 1.13 until they fix it?